### PR TITLE
Benchmarking w/ ASV + Github Actions

### PR DESCRIPTION
**Pull request by thomasaarholt**
**TLDR:** Monitor hyperspy code speed through writing benchmarks. Automatically generated. [Commits have a GUI](https://thomasaarholt.github.io/hyperspy-bench/) and [PRs print text](https://github.com/thomasaarholt/hyperspy/runs/970602990?check_suite_focus=true) (click "Compare Benchmarks").

### Benchmarking with AirSpeed Velocity (ASV) and GitHub Actions
I've written several Github Actions that use [ASV](https://github.com/airspeed-velocity/asv) to test HyperSpy PRs and commits to monitor regressions and speedups of the HyperSpy code. The commits are monitored through a separate repo at `hyperspy/hyperspy-benchmark`, branch `gh-pages` (which GitHub automatically hosts as a [website](https://thomasaarholt.github.io/hyperspy-bench/)). It currently exists at [`thomasaarholt/hyperspy-bench`](https://github.com/thomasaarholt/hyperspy-bench/tree/gh-pages) and will be transferred once people have had a look. Here the commit benchmark `results` and `html` are stored and the [website is published](https://thomasaarholt.github.io/hyperspy-bench/).

Two Actions will run automatically:
- The PR Action (`bench_PR.yml`) compares the latest commit in a PR to the latest `RELEASE_next_minor` and can be read under `Checks`->`Benchmark PR vs Rnm`->`Ubuntu/Windows`->`Compare Benchmarks`. Here the benchmarks are listed in terms of time spent and ratio between the PR and upstream.
- The commit Action (`bench_commits.yml`) monitors pushes to RELEASE_next_minor. It will run the benchmarks on the commit, add the results to the `hyperspy-benchmark` repo, and update the html in the same repo.

Two Actions can be run manually:
- The Publish Action (`bench_publish.yml`) will generate html from all results present in the hyperspy-branch repo. Good in case the website breaks for some reason.
- The Manual Range Action (`bench_manual_range.yml` aka "400 commits") will benchmark a number of commit SHAs over a range from start to finish. This is useful to add history on the benchmark website.

I have used the Range Action to populate the website with 400 commits. On four benchmarks, these 400 commits took about two hours per OS (Ubuntu and Windows). [Take a look](https://thomasaarholt.github.io/hyperspy-bench/). Try clicking on a graph. The PR action runs in ~7 minutes. 

See #2469 to suggest specific things to benchmark.

### Things to change to move to hyperspy/hyperspy-benchmark
- [ ] Create hyperspy/hyperspy-bench and push `asv.conf.json` to it from the `thomasaarholt` repo.
- [ ] Generate a [Personal Access Token](https://github.com/settings/tokens) and add it to hyperspy -> Settings -> Secrets so the action has push-access to `/hyperspy-benchmark`. Add as "BENCHMARK_PAT".
- [ ] Run the `Range Action` to populate it with historical commits.

I am happy to do the above once we are agreed on any outstanding discussion points.

### Summary of details about workflow
The general idea is to not push anything to the `hyperspy/hyperspy` repo. The config (`asv.conf.json`) is kept on hyperspy-benchmark(could consider keeping a copy on `/hyperspy` too). It, and any previous results, are copied into a local checkout of the PR/commit branch. The benchmarks are run, the website generated, and all results and website are pushed to hyperspy-branch. During the PR Action, no changes are made to hyperspy-bench.

### CODE REVIEW

**Comment by ericpre** [comment]
This looks promising (and much closer to what I was thinking), a few comments:
- I don't think that this is necessary to have any configuration file in the `hyperspy/hyperspy` repository: a daily cron job should be enough. This would simplify setting up/maintenance and there will not be any need to add secrets to the repository.
- for the repository name, I think `hyperspy-benchmark` is more clear.
- your repository can be transferred to the hyperspy organisation as it is and then we just need to set up the github pages (no need to populate again the history).

**Comment by ericpre** [comment]
https://napari.org/docs/developers/BENCHMARKS.html

**Comment by ericpre** [comment]
@thomasaarholt, what are your plans to finish this?

**Comment by thomasaarholt** [comment]
Aiming to get this in this week / Tuesday. To sum up where we left it from your comments:
>- I don't think that this is necessary to have any configuration file in the hyperspy/hyperspy repository: a daily cron job should be enough. This would simplify setting up/maintenance and there will not be any need to add secrets to the repository.

Do you mean that we only benchmark once, daily? This means we are not benchmarking PRs before they are merged. That's fine though, it can be a bit much under "Checks" on a PR anyway. I should be able to set up the daily cron job as a github action. 

>- for the repository name, I think hyperspy-benchmark is more clear.

agreed

>- your repository can be transferred to the hyperspy organisation as it is and then we just need to set up the github pages (no need to populate again the history).

In summary:

I'll create `hyperspy/hyperspy-benchmark`, create branch `gh-pages`, copy over the existing `thomasaarholt/hyperspy-bench` repo. Then set up a github action / cron job to benchmark at some time during the night. I'll prevent the job from running if the latest commit has already been benchmarked.

**Comment by ericpre** [comment]
> Do you mean that we only benchmark once, daily?

Yes, this is what I meant.

> In summary:
>
> I'll create hyperspy/hyperspy-benchmark, create branch gh-pages, copy over the existing thomasaarholt/hyperspy-bench repo.
Then set up a github action / cron job to benchmark at some time during the night. I'll prevent the job from running if the latest commit has already been benchmarked.

Would it be more simple to transfer the repository as mentioned above?

**Comment by ericpre** [comment]
https://github.com/matplotlib/mpl-bench
